### PR TITLE
[V6] Register OctaneReloadPermissions listener for Laravel Octane

### LINKED ISSUE #2401: Wrong permissions on refresh (Vapor/Octane)
When creating a new role with permissions and assign it to a user, the user randomly gets a 403 message.

The issue seems that i'm using Octane on Laravel Vapor. When you refresh your screen or navigate to another screen you might end up on a different lambda server which still has old cached permissions. after a re-deploy it works fine. So a lamba server must have the permissions still in the memory.

Laravel version: v10.7.1
Permission version: v5.10.0
PHP version: 8.1
Database version: MySQL 8.0.32

**To Reproduce**
Create a new role with permissions and assign it to a user, login as that user.

### PR DESCRIPTION
**Pull request by erikn69**
i'm not 100% sure on this 
i can't test octane 

this was a test for octane users to try but no one gave me feedback that it works

On https://github.com/spatie/laravel-permission/issues/2322#issuecomment-1493242776 was said that the problem was the **spatie/once** package, and that **spatie/laravel-permission** works correctly on octane

**I recommend not merging it until you do more testing that it's the correct approach.**

### CODE REVIEW

**Comment by drbyte** [comment]
> I recommend not merging it until you do more testing that it's the correct approach

Agreed. I more just wanted it here as an easy place to reference the specific proposed fix. And if it's deemed "right" and not problematic, it's a simple thing then to merge it.
Thanks!

**Comment by drbyte** [comment]
@sweebee Are you able to test this? In reference to your issue #2401

**Comment by sweebee** [comment]
@drbyte Yes, its working!

**Comment by davidjr82** [comment]
This solution will reload the permissions in every single tick of the octane server. Wouldn't this result in a drop of performance?

Also, we should make (I don't know how...) a test case where we can see it failing without this change, otherwise it wouldn't add anything.

**Comment by drbyte** [comment]
> This solution will reload the permissions in every single tick of the octane server. Wouldn't this result in a drop of performance?

@sweebee Do you mind testing what benefits or drawbacks occur if not registering it with the Tick events?

> Also, we should make (I don't know how...) a test case where we can see it failing without this change, otherwise it wouldn't add anything.

@sweebee Any thoughts on how best to automate such tests?

File: src/PermissionServiceProvider.php [hunk 1]
```diff
@@ -86,6 +90,19 @@ protected function registerCommands(): void
...
 
+    protected function registerOctaneListener(): void
+    {
+        if ($this->app->runningInConsole() || ! $this->app['config']->get('octane.listeners')) {
+            return;
+        }
+
+        $dispatcher = $this->app[Dispatcher::class];
+
+        $dispatcher->listen(\Laravel\Octane\Events\RequestReceived::class, OctaneReloadPermissions::class);
```
Comment (drbyte) [inline src/PermissionServiceProvider.php:101]: Perhaps we should just listen on `OperationTerminated`?

> The `OperationTerminated` event is a interface that is implemented by `TickTerminated`, `TaskTerminated` and `RequestTerminated`, which all get fired depending on the context of the application (receiving a request or executing a job/command).

Comment (erikn69) [reply src/PermissionServiceProvider.php]: Done

Comment (drbyte) [reply src/PermissionServiceProvider.php]: Thanks

**Comment by drbyte** [comment]
@sweebee @davidjr82 does the change to listening on `OperationTerminated` work satisfactorily in your app?

**Comment by davidjr82** [comment]
For me the issue was in a different package (spatie/once) so this package worked without any change.

**Comment by davidjr82** [comment]
> >this package worked without any change.
> > this result in a drop of performance
> 
> it would be better to make it optional, so, can be activated from permission.php for whoever needs it
> Like 
> ```php
> /*
>  * When set to true, the Spatie\Permission\Listeners\OctaneReloadPermissions listener will be registered
>  * on the Laravel\Octane\Events\OperationTerminated event, this will refresh permissions on every
>  * TickTerminated, TaskTerminated and RequestTerminated
>  */
> 'register_octane_permissions_refresh' => true,
> ```

I agree on this, although I wouldn't push a feature without a working test. 

But if in this case we don't know how to do it, at least make it optimal sounds reasonable to me.

**Comment by erikn69** [comment]
> it would be better to make it optional

that could also apply

**Comment by drbyte** [comment]
I'm in agreement with adding a config switch. Just undecided about the best name for it. 
The variant I'm contemplating is `register_octane_reset_listener`

**Comment by drbyte** [comment]
> adding tests

I agree: it would be preferable to get some tests set up for Octane.  **#HelpWanted**

**Comment by drbyte** [comment]
@freekmurze do you have a preference on incorporating tests for this?

**Comment by jihadismail8** [comment]
for some reason Spatie Permissions , gives me an error of not enough permissions after ~14 requests in a row  under Octane RoadRunner !!  ""message": "User does not have the right permissions. Necessary permissions are Manage Permissions","  , even thought the 13 requests where fine , then sudden not enough permission , tried  almost everything , and i have the last version of laravel 11 , fresh spatie Permission ....  here is the trace " "trace": [
        {
            "file": "/var/www/html/vendor/spatie/laravel-permission/src/Middleware/PermissionMiddleware.php",
            "line": 36,
            "function": "forPermissions",
            "class": "Spatie\\Permission\\Exceptions\\UnauthorizedException",
            "type": "::"
        },
        {
            "file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php",
            "line": 183,
            "function": "handle",
            "class": "Spatie\\Permission\\Middleware\\PermissionMiddleware",
            "type": "->"
        },
        {
            "file": "/var/www/html/app/Http/Middleware/Cors.php",
            "line": 23,
            "function": "Illuminate\\Pipeline\\{closure}",
            "class": "Illuminate\\Pipeline\\Pipeline",
            "type": "->"
        },
        {
            "file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php",
            "line": 183,
            "function": "handle",
            "class": "App\\Http\\Middleware\\Cors",
            "type": "->"
        },
        {
            "file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Routing/Middleware/SubstituteBindings.php",
            "line": 50,
            "function": "Illuminate\\Pipeline\\{closure}",
            "class": "Illuminate\\Pipeline\\Pipeline",
            "type": "->"
        },
        {
            "file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php",
            "line": 183,
            "function": "handle",
            "class": "Illuminate\\Routing\\Middleware\\SubstituteBindings",
            "type": "->"
        },
        {
            "file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Auth/Middleware/Authenticate.php",
            "line": 64,
            "function": "Illuminate\\Pipeline\\{closure}",
            "class": "Illuminate\\Pipeline\\Pipeline",
            "type": "->"
        },
        {
            "file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php",
            "line": 183,
            "function": "handle",
            "class": "Illuminate\\Auth\\Middleware\\Authenticate",
            "type": "->"
        },
        {
            "file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php",
            "line": 119,
            "function": "Illuminate\\Pipeline\\{closure}",
            "class": "Illuminate\\Pipeline\\Pipeline",
            "type": "->"
        },
        {
            "file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Routing/Router.php",
            "line": 805,
            "function": "then",
            "class": "Illuminate\\Pipeline\\Pipeline",
            "type": "->"
        },
        {
            "file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Routing/Router.php",
            "line": 784,
            "function": "runRouteWithinStack",
            "class": "Illuminate\\Routing\\Router",
            "type": "->"
        },
        {
            "file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Routing/Router.php",
            "line": 748,
            "function": "runRoute",
            "class": "Illuminate\\Routing\\Router",
            "type": "->"
        },
        {
            "file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Routing/Router.php",
            "line": 737,
            "function": "dispatchToRoute",
            "class": "Illuminate\\Routing\\Router",
            "type": "->"
        },
        {
            "file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php",
            "line": 200,
            "function": "dispatch",
            "class": "Illuminate\\Routing\\Router",
            "type": "->"
        },
        {
            "file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php",
            "line": 144,
            "function": "Illuminate\\Foundation\\Http\\{closure}",
            "class": "Illuminate\\Foundation\\Http\\Kernel",
            "type": "->"
        },
        {
            "file": "/var/www/html/vendor/livewire/livewire/src/Features/SupportDisablingBackButtonCache/DisableBackButtonCacheMiddleware.php",
            "line": 19,
            "function": "Illuminate\\Pipeline\\{closure}",
            "class": "Illuminate\\Pipeline\\Pipeline",
            "type": "->"
        },
        {
            "file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php",
            "line": 183,
            "function": "handle",
            "class": "Livewire\\Features\\SupportDisablingBackButtonCache\\DisableBackButtonCacheMiddleware",
            "type": "->"
        },
        {
            "file": "/var/www/html/vendor/barryvdh/laravel-debugbar/src/Middleware/InjectDebugbar.php",
            "line": 66,
            "function": "Illuminate\\Pipeline\\{closure}",
            "class": "Illuminate\\Pipeline\\Pipeline",
            "type": "->"
        },
        {
            "file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php",
            "line": 183,
            "function": "handle",
            "class": "Barryvdh\\Debugbar\\Middleware\\InjectDebugbar",
            "type": "->"
        },
        {
            "file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/TransformsRequest.php",
            "line": 21,
            "function": "Illuminate\\Pipeline\\{closure}",
            "class": "Illuminate\\Pipeline\\Pipeline",
            "type": "->"
        },
        {
            "file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/ConvertEmptyStringsToNull.php",
            "line": 31,
            "function": "handle",
            "class": "Illuminate\\Foundation\\Http\\Middleware\\TransformsRequest",
            "type": "->"
        },
        {
            "file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php",
            "line": 183,
            "function": "handle",
            "class": "Illuminate\\Foundation\\Http\\Middleware\\ConvertEmptyStringsToNull",
            "type": "->"
        },
        {
            "file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/TransformsRequest.php",
            "line": 21,
            "function": "Illuminate\\Pipeline\\{closure}",
            "class": "Illuminate\\Pipeline\\Pipeline",
            "type": "->"
        },
        {
            "file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/TrimStrings.php",
            "line": 51,
            "function": "handle",
            "class": "Illuminate\\Foundation\\Http\\Middleware\\TransformsRequest",
            "type": "->"
        },
        {
            "file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php",
            "line": 183,
            "function": "handle",
            "class": "Illuminate\\Foundation\\Http\\Middleware\\TrimStrings",
            "type": "->"
        },
        {
            "file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Http/Middleware/ValidatePostSize.php",
            "line": 27,
            "function": "Illuminate\\Pipeline\\{closure}",
            "class": "Illuminate\\Pipeline\\Pipeline",
            "type": "->"
        },
        {
            "file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php",
            "line": 183,
            "function": "handle",
            "class": "Illuminate\\Http\\Middleware\\ValidatePostSize",
            "type": "->"
        },
        {
            "file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/PreventRequestsDuringMaintenance.php",
            "line": 110,
            "function": "Illuminate\\Pipeline\\{closure}",
            "class": "Illuminate\\Pipeline\\Pipeline",
            "type": "->"
        },
        {
            "file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php",
            "line": 183,
            "function": "handle",
            "class": "Illuminate\\Foundation\\Http\\Middleware\\PreventRequestsDuringMaintenance",
            "type": "->"
        },
        {
            "file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Http/Middleware/HandleCors.php",
            "line": 62,
            "function": "Illuminate\\Pipeline\\{closure}",
            "class": "Illuminate\\Pipeline\\Pipeline",
            "type": "->"
        },
        {
            "file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php",
            "line": 183,
            "function": "handle",
            "class": "Illuminate\\Http\\Middleware\\HandleCors",
            "type": "->"
        },
        {
            "file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Http/Middleware/TrustProxies.php",
            "line": 57,
            "function": "Illuminate\\Pipeline\\{closure}",
            "class": "Illuminate\\Pipeline\\Pipeline",
            "type": "->"
        },
        {
            "file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php",
            "line": 183,
            "function": "handle",
            "class": "Illuminate\\Http\\Middleware\\TrustProxies",
            "type": "->"
        },
        {
            "file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php",
            "line": 119,
            "function": "Illuminate\\Pipeline\\{closure}",
            "class": "Illuminate\\Pipeline\\Pipeline",
            "type": "->"
        },
        {
            "file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php",
            "line": 175,
            "function": "then",
            "class": "Illuminate\\Pipeline\\Pipeline",
            "type": "->"
        },
        {
            "file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php",
            "line": 144,
            "function": "sendRequestThroughRouter",
            "class": "Illuminate\\Foundation\\Http\\Kernel",
            "type": "->"
        },
        {
            "file": "/var/www/html/vendor/laravel/octane/src/ApplicationGateway.php",
            "line": 36,
            "function": "handle",
            "class": "Illuminate\\Foundation\\Http\\Kernel",
            "type": "->"
        },
        {
            "file": "/var/www/html/vendor/laravel/octane/src/Worker.php",
            "line": 84,
            "function": "handle",
            "class": "Laravel\\Octane\\ApplicationGateway",
            "type": "->"
        },
        {
            "file": "/var/www/html/vendor/laravel/octane/bin/roadrunner-worker",
            "line": 54,
            "function": "handle",
            "class": "Laravel\\Octane\\Worker",
            "type": "->"
        },
        {
            "file": "/var/www/html/vendor/bin/roadrunner-worker",
            "line": 119,
            "function": "include"
        } "

**Comment by jihadismail8** [comment]
forgot to add that in php artisan serve , it works fine

**Comment by jihadismail8** [comment]
ok found the problem ,  it was because of teams , since teams is being flushed the Users loses permissions
that creates a dilemma now

**Pull request merged**